### PR TITLE
SASL Authentication added

### PR DESCRIPTION
**Pull request by gstark0**
Added new module called sasl.py, to authorize identity using SASL. You should register Begiak on NickServ first (if not done before). Also, there are few changes in config file, so if you run on already existing instance of Begiak, this changes should be made to default.py:
- Add "sasl = False(or True)" line.
- Uncomment "password = 'example'" line, and enter password to NickServ.
- Remove empty string in "ignore = ['']", line. So it would be "ignore = []" now.
screen shot 2018-01-05 at 17 12 05

### CODE REVIEW

File: bot.py [hunk 1]
```diff
@@ -236,7 +236,7 @@ def limit(self, origin, func):
         return False
 
     def dispatch(self, origin, args): 
-        bytes, event, args = args[0], args[1], args[2:]
+        bytes, event, args = args[0], args[1], args
```
Comment (sushain97) [inline bot.py]: These lines don't do the same thing.

File: default.py.example [hunk 2]
```diff
@@ -2,13 +2,14 @@ nick = 'phenny'
...
 ssl = False
+sasl = False
 ipv6 = False
 channels = ['#example', '#test']
 owner = 'yournickname'
 greetings = {'#example': 'example greeting %name %channel', '#test': 'welcome to %channel %name'}
 
 # password is the NickServ password, serverpass is the server password
-# password = 'example'
+password = 'example'
```
Comment (sushain97) [inline default.py.example]: We should re-comment this.

File: default.py.example [hunk 3]
```diff
@@ -2,13 +2,14 @@ nick = 'phenny'
 host = 'irc.example.net'
 port = 6667
 ssl = False
+sasl = False
```
Comment (sushain97) [inline default.py.example]: Maybe mention here what other options this requires.

File: modules/sasl.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
+#!/usr/bin/env python3
+"""
+sasl.py - phenny SASL Authentication module
+"""
+import base64
+import logging
+
+logger = logging.getLogger('phenny')
+
+def irc_cap (phenny, input):
```
Comment (sushain97) [inline modules/sasl.py]: Spacing

File: modules/sasl.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+	else:
+		irc_cap_end(phenny, input)
+
+	return
+irc_cap.rule = r'(.*)'
+irc_cap.event = 'CAP'
+irc_cap.priority = 'high'
+
+
+def irc_authenticated (phenny, input):
```
Comment (sushain97) [inline modules/sasl.py]: Spacing

File: modules/sasl.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+		irc_cap_end()
+		return
+
+	while len(auth) >= 400:
+		out = auth[0:400]
+		auth = auth[401:]
+		phenny.write(('AUTHENTICATE', out))
+	phenny.write(('AUTHENTICATE', auth))
+
+	return
```
Comment (sushain97) [inline modules/sasl.py]: Superfluous

File: modules/sasl.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+irc_authenticated.rule = r'(.*)'
+irc_authenticated.event = 'AUTHENTICATE'
+irc_authenticated.priority = 'high'
+
+
+def irc_903 (phenny, input):
```
Comment (sushain97) [inline modules/sasl.py]: Spacing and unnecessary return on all these functions :)

**Comment by gstark0** [comment]
@sushain97 Fixed :)

**Comment by sushain97** [comment]
We're still setting `args = args`?

File: modules/sasl.py [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+
+def irc_905(phenny, input):
+	irc_cap_end(phenny, input)
+	logger.error('SASL Authentication failed: Message too long (up to 400 bytes). ')
+	return
+irc_905.rule = r'(.*)'
+irc_905.event = '905'
+irc_905.priority = 'high'
+
+def irc_906 (phenny, input):
```
Comment (sushain97) [inline modules/sasl.py]: I think I commented on the spacing on this function and the rest of them? There shouldn't be space between the function name and opening paren. Also, none of these functions need a `return`.

**Comment by gstark0** [comment]
@sushain97 sorry for that spaces, I removed some of them and forgot about the rest :D

**Comment by sushain97** [comment]
@jonorthwash could you possibly test this?

**Comment by jonorthwash** [comment]
Now begiak is registered with NickServ, but we need to test a bit, I guess?

**Pull request merged**